### PR TITLE
clustermesh-apiserver: Document and warn against potential connectivity issue

### PR DESCRIPTION
**Pull request by qmonnet**
- **docs: List clustermesh-apiserver limitation in troubleshooting section**
- **helm: Add check for clustermesh-apiserver + KPR/tunnel/WireGuard/HFW**

Relates to: https://github.com/cilium/cilium/issues/31209

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (giorio94) [inline Documentation/operations/troubleshooting_clustermesh.rst]: I wouldn't personally mention the configuration timing as a requisite here, as we cannot exclude that the same issue could occur after the initial configuration phase, e.g., in case of node churn. In that case, depending again on timing, the Cilium agent on the new node may not be able to connect to the remote clustermesh-apiserver if the agents in the other cluster already discovered about that node, and started encrypting traffic towards it.

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (giorio94) [inline Documentation/operations/troubleshooting_clustermesh.rst:178]: I wonder if it could be worth mentioning that the very same issue happens with WireGuard node to node encryption enabled (regardless of whether host firewall is enabled).

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (giorio94) [inline Documentation/operations/troubleshooting_clustermesh.rst]: I don't think that KVStoreMesh plays a role here, as in any case one component (the kvstoremesh container in that case) would still need to be able to connect to the remote clustermesh-apiserver service. I indeed managed to reproduce the same issue with KVStoreMesh enabled.

**Review by nathanjsweet — APPROVED** [review]

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (giorio94) [inline Documentation/operations/troubleshooting_clustermesh.rst]: I managed to reproduce a sort of similar issue even with KPR disabled, if kvstoremesh is enabled (i.e., traffic is pod to nodeport, rather than node to nodeport). My guess is that the latter case is instead handled by [this logic](https://github.com/cilium/cilium/blob/43642e5cda6eeb8c7ab7af9d64579ce725501e85/bpf/bpf_lxc.c#L1209-L1211), which forwards the nodeport reply to the stack, and then back via the native device.

Let's consider a scenario in which cluster A connected to cluster B, but not vice versa (e.g., commenting out the last step from your patch). The initial connection from kvstoremesh on A to the clustermesh-apiserver on B works because neither hosting agents knows about the nodes in the other cluster. However, after the initial connection, A learns about the nodes in B, and subsequent connections to the clustermesh-apiserver on B will go through the tunnel, and get dropped on receive side due to the missing wireguard keys. 

This can be verified running `kubectl exec -t -n kube-system deploy/clustermesh-apiserver -c kvstoremesh -- clustermesh-apiserver kvstoremesh-dbg troubleshoot` on A, which fails, while `...  kvstoremesh-dbg status` reports success.

Although the situation can converge if B connects back to A before that the initial TCP connection gets closed, this is still risky. Hence, I'd suggest to update the docs to not mention KPR as one of the requisites for this issue to happen, considering that even KPR disabled can still trigger issues in certain conditions.

**Review by giorio94 — CHANGES_REQUESTED** [review]
Thanks Quentin! I've left a couple of comments inline about the possibility of removing some of the conditions from the list, as I managed to reproduce the issue in other cases as well. The helm validation should be also updated accordingly.

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (qmonnet) [reply Documentation/operations/troubleshooting_clustermesh.rst]: OK, that makes sense. I'll drop this item from the list.

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (qmonnet) [reply Documentation/operations/troubleshooting_clustermesh.rst:178]: Let's mention it in that case. I'll add the note to the WireGuard page as well, since we're at it

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (qmonnet) [reply Documentation/operations/troubleshooting_clustermesh.rst]: Right, I got confused about that one, sorry. I'll remove it.

File: Documentation/operations/troubleshooting_clustermesh.rst
Comment (qmonnet) [reply Documentation/operations/troubleshooting_clustermesh.rst]: Nice find! OK, I'll drop the KPR “prerequisite” to trigger the issue (and from the Helm validation).

I'll also report your comment to the original issue, so we can better remember it.

**Comment by qmonnet** [comment]
I adjusted the docs and validation, thanks a lot Marco!

**Review by giorio94 — APPROVED** [review]
Thanks!

**Review by brb — APPROVED** [review]

**Comment by giorio94** [comment]
I'm marking this PR as `dont-merge/discussion`, as I'd like to first perform a couple of final tests with the service of type LoadBalancer.

**Comment by giorio94** [comment]
> I'm marking this PR as `dont-merge/discussion`, as I'd like to first perform a couple of final tests with the service of type LoadBalancer.

Using a LoadBalancer service turned out not addressing the issue in all cases: https://github.com/cilium/cilium/issues/31209#issuecomment-2196545466

**Review by sayboras — APPROVED** [review]

**Review by borkmann — APPROVED** [review]
Changes lgtm, modulo missing mention of how a resolution would look like, but as I understand this is still tbd.

**Comment by julianwiedmann** [comment]
@giorio94 is this still blocked on discussion? :)

**Comment by giorio94** [comment]
> @giorio94 is this still blocked on discussion? :)

I'm not sure if there's been much progress recently (not on my side), but using a LoadBalancer service (as I initially supposed, and as documented here) does not actually prevent this issue from occurring: https://github.com/cilium/cilium/issues/31209#issuecomment-2196545466.

**Pull request closed without merging**
